### PR TITLE
Support better matchers for guardduty

### PR DESCRIPTION
**Pull request by ajvb**
Add GuardDutyConfig and GuardDutyFindingMatcher which add support for a
json file that contains all "ignore" and "high severity" matchers (and
everything else is considered "low severity"). These matchers currently
support matching on finding type, account id, resource tags, and domain
name within finding (specific to DNS_REQUEST related findings).

Example config file:
```json
{
  "ignore_matchers": [
    {"aws_tags": {"Name": "CloudGoat.+"}},
    {"finding_type": "Recon:EC2.+"},
    {"finding_type": "Trojan.+", "domain_name": "ignore.com"}
  ],
  "high_severity_matchers": [
    {"finding_type": "Trojan.+"},
    {"finding_type": "Backdoor.+"},
    {
      "account_id": "999999999",
      "finding_type": "Stealth:IAMUser/LoggingConfigurationModified"
    }
  ]
}
```

### CODE REVIEW

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyConfig.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.mozilla.secops.GcsUtil;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+
+/* Class for configuration around our use of Guardduty, i.e. which Finding's to ignore / escalate */
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyConfig.java:12]: `/**`

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyConfig.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  public List<GuardDutyFindingMatcher> getIgnoreMatchers() {
+    return ignoreMatchers;
+  }
+
+  @JsonProperty("high_severity_matchers")
+  public List<GuardDutyFindingMatcher> getHighSeverityMatchers() {
+    return highSeverityMatchers;
+  }
+
+  /* Create a new empty GuardDutyConfig */
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyConfig.java:53]: `/**`

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyConfig.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    }
+    if (in == null) {
+      throw new IOException("guardduty config resource not found");
+    }
+    ObjectMapper mapper = new ObjectMapper();
+    return mapper.readValue(in, GuardDutyConfig.class);
+  }
+
+  @JsonProperty("ignore_matchers")
+  public List<GuardDutyFindingMatcher> getIgnoreMatchers() {
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyConfig.java:44]: Missing javadoc

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyConfig.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    return ignoreMatchers;
+  }
+
+  @JsonProperty("high_severity_matchers")
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyConfig.java:48]: Missing javadoc

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+import com.amazonaws.services.guardduty.model.Resource;
+import com.amazonaws.services.guardduty.model.Tag;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+/* Class for matching against Guardduty findings, used by Gatekeeper to bucket findings into different categories */
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java:15]: `/**`

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+  private String findingType;
+  private Pattern findingTypeMatcher;
+  private Map<String, String> awsTags;
+  private Map<String, Pattern> awsTagsPatterned;
+  private String accountId;
+  private String domainName;
+
+  @JsonProperty("finding_type")
+  public void setFindingType(String findingType) {
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java:27]: Missing javadoc

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    this.findingType = findingType;
+    if (findingType != null) {
+      findingTypeMatcher = Pattern.compile(findingType);
+    }
+  }
+
+  @JsonProperty("aws_tags")
+  public void setAwsTags(Map<String, String> awsTags) {
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java:35]: Missing javadoc

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    this.awsTags = awsTags;
+    awsTagsPatterned = new HashMap<String, Pattern>();
+    for (Map.Entry<String, String> tag : awsTags.entrySet()) {
+      awsTagsPatterned.put(tag.getKey(), Pattern.compile(tag.getValue()));
+    }
+  }
+
+  @JsonProperty("account_id")
+  public void setAccountId(String accountId) {
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java:44]: Missing javadoc

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    this.accountId = accountId;
+  }
+
+  @JsonProperty("domain_name")
+  public void setDomainName(String domainName) {
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java:49]: Missing javadoc

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    this.domainName = domainName;
+  }
+
+  public String getFindingType() {
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java:53]: Missing javadoc

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    return findingType;
+  }
+
+  public Map<String, String> getAwsTags() {
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java:57]: Missing javadoc

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    return awsTags;
+  }
+
+  public String getAccountId() {
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java:61]: Missing javadoc

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+      if (findingTypeMatcher.matcher(f.getType()).matches()) {
+        ftMatches = true;
+      }
+    } else {
+      ftMatches = true;
+    }
+
+    // Check account id
+    if (accountId != null) {
+      if (f.getAccountId() != null && f.getAccountId().equals(accountId)) {
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java:92]: If `f.getAccountId()` returns `null` should this return false right away like the previous block?

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+      idMatches = true;
+    }
+
+    // Check resource tags
+    if (awsTags != null) {
+      Resource rsrc = f.getResource();
+      if (rsrc != null) {
+        InstanceDetails idets = rsrc.getInstanceDetails();
+        if (idets != null) {
+          List<Tag> tags = idets.getTags();
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java:105]: Similar question to previous, should we just return here if we don't have tags but need them?

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java (hunk 13 above)
Comment (ajvb) [reply src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java]: Yeah absolutely. Fixed

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java (hunk 14 above)
Comment (ajvb) [reply src/main/java/com/mozilla/secops/gatekeeper/GuardDutyFindingMatcher.java]: Yeah, same here, fixed.

**Pull request merged**